### ISSUE TITLE
Kubectl client Kubectl logs should be able to retrieve and filter logs [Conformance] failing on all 1.2 branch jobs

### ISSUE THREAD

**Issue by jlowdermilk** [kubernetes/kubernetes#25426]
Example from [kubernetes-e2e-gce-release-1.2/2834](https://console.cloud.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-release-1.2/2834) ([jenkins page](http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-release-1.2/2834/consoleFull))

```
19:04:56 STEP: Destroying namespace "e2e-tests-kubectl-6sila" for this suite.
19:04:56 
19:04:56 
19:04:56 • Failure [59.548 seconds]
19:04:56 Kubectl client
19:04:56 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:1164
19:04:56   Kubectl logs
19:04:56   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:857
19:04:56     should be able to retrieve and filter logs [Conformance] [It]
19:04:56     /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:856
19:04:56 
19:04:56     Expected
19:04:56         <int>: 0
19:04:56     to equal
19:04:56         <int>: 1
19:04:56 
19:04:56     /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:831
```

It looks like the test started failing consistently yesterday at 7pm. cc @k8s-oncall

**Comment by smarterclayton** [comment]
Did the redis master image change?

**Comment by smarterclayton** [comment]
This started breaking in OpenShift tests (which is on a ~3 week old Kube master commit), so I don't think it's a change in Kube.

**Comment by deads2k** [comment]
> This started breaking in OpenShift tests (which is on a ~3 week old Kube master commit), so I don't think it's a change in Kube.

@kargakis any chance this is caused by https://github.com/kubernetes/kubernetes/pull/23590 ?  

@smarterclayton we picked that back into origin yesterday.

**Comment by kargakis** [comment]
> @kargakis any chance this is caused by #23590 ?

Hm, had a look but that [test](https://github.com/kubernetes/kubernetes/blob/13f60ea8620bbc66727dd98e7e0b6ec71d0465fa/test/e2e/kubectl.go#L869-L879) seems to run `kubectl logs` against a pod name while 23590 is about resolving higher level resources to pod names, ie. different path.

**Comment by jlowdermilk** [comment]
@smarterclayton, looks like that's what happened. On master we're using a tagged redis image, which is why this is only failing on 1.2 branch tests. Cherry-picking #24849 should fix.

**Comment by jlowdermilk** [comment]
Redis image was a red herring. The test is failing because `kubectl logs --limit-bytes=1` isn't returning any output when run from the test. This is odd because it works just fine when you run it standalone (same version). Still digging.

**Comment by jlowdermilk** [comment]
Ok I think I found the issue. The test should actually be failing at HEAD, but it's being silently skipped because of a broken version check.

```
16:22:11 [It] should be able to retrieve and filter logs [Conformance]
16:22:11   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:901
16:22:11 May 11 16:22:03.946: INFO: Selector matched 1 pods for map[app:redis]
16:22:11 May 11 16:22:03.946: INFO: ForEach: Found 0 pods from the filter.  Now looping through them.
16:22:11 [AfterEach] [k8s.io] Kubectl logs
16:22:11   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:858
16:22:11 STEP: using delete to clean up resources
16:22:11 May 11 16:22:03.946: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.197.36.252 --kubeconfig=/workspace/.kube/config delete --grace-period=0 -f /workspace/kubernetes/test/e2e/testing-manifests/kubectl/redis-master-controller.json --namespace=e2e-tests-kubectl-i75d1'
16:22:11 May 11 16:22:06.156: INFO: stderr: ""
16:22:11 May 11 16:22:06.156: INFO: stdout: "replicationcontroller \"redis-master\" deleted"
16:22:11 May 11 16:22:06.156: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.197.36.252 --kubeconfig=/workspace/.kube/config get rc,svc -l name=nginx --no-headers --namespace=e2e-tests-kubectl-i75d1'
16:22:11 May 11 16:22:06.235: INFO: stderr: ""
16:22:11 May 11 16:22:06.235: INFO: stdout: ""
16:22:11 May 11 16:22:06.235: INFO: Running '/workspace/kubernetes/platforms/linux/amd64/kubectl --server=https://104.197.36.252 --kubeconfig=/workspace/.kube/config get pods -l name=nginx --namespace=e2e-tests-kubectl-i75d1 -o go-template={{ range .items }}{{ if not .metadata.deletionTimestamp }}{{ .metadata.name }}{{ "\n" }}{{ end }}{{ end }}'
16:22:11 May 11 16:22:06.311: INFO: stderr: ""
16:22:11 May 11 16:22:06.312: INFO: stdout: ""
16:22:11 [AfterEach] [k8s.io] Kubectl client
16:22:11   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:109
16:22:11 May 11 16:22:06.312: INFO: Waiting up to 1m0s for all nodes to be ready
16:22:11 STEP: Destroying namespace "e2e-tests-kubectl-i75d1" for this suite.
16:22:11 
16:22:11 
16:22:11 • [SLOW TEST:7.605 seconds]
16:22:11 [k8s.io] Kubectl client
16:22:11 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:505
16:22:11   [k8s.io] Kubectl logs
16:22:11   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:505
16:22:11     should be able to retrieve and filter logs [Conformance]
16:22:11     /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:901
16:22:11 ------------------------------
```

The reason the test is failing is because the [`Exec()`](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L1554) command strips whitespace. Since log output from the redis-master container looks like
```
                _._                                                  
           _.-``__ ''-._                                             
      _.-``    `.  `_.  ''-._           Redis 3.2.0 (00000000/0) 64 bit
  .-`` .-```.  ```\/    _.,_ ''-._                                   
 (    '      ,       .-`  | `,    )     Running in standalone mode
 |`-._`-...-` __...-.``-._|'` _.-'|     Port: 6379
 |    `-._   `._    /     _.-'    |     PID: 1
  `-._    `-._  `-./  _.-'    _.-'                                   
 |`-._`-._    `-.__.-'    _.-'_.-'|                                  
 |    `-._`-._        _.-'_.-'    |           http://redis.io        
  `-._    `-._`-.__.-'_.-'    _.-'                                   
 |`-._`-._    `-.__.-'    _.-'_.-'|                                  
 |    `-._`-._        _.-'_.-'    |                                  
  `-._    `-._`-.__.-'_.-'    _.-'                                   
      `-._    `-.__.-'    _.-'                                       
          `-._        _.-'                                           
              `-.__.-'                                               

1:M 11 May 23:49:34.199 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
1:M 11 May 23:49:34.200 # Server started, Redis version 3.2.0
1:M 11 May 23:49:34.200 # WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo never > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled.
1:M 11 May 23:49:34.200 * The server is now ready to accept connections on port 6379
```
The first few bytes are whitespace, and thus the command returns an empty string. Fix is to remove the whitespace stripping which should really have been done a long time ago (as the TODO suggests).

The much more concerning issue is that tests are getting skipped and look like they are passing. Filed #25515 to look at spuriously passing tests.

**Comment by jlowdermilk** [comment]
Hasty conclusion #2 also wrong. The version check is correct, but `forEachPod` was called before waiting for pods to be scheduled, so the whole test is no-op.
```
16:22:11 May 11 16:22:03.946: INFO: Selector matched 1 pods for map[app:redis]
16:22:11 May 11 16:22:03.946: INFO: ForEach: Found 0 pods from the filter.  Now looping through them.
```
There's a couple other tests that are spuriously passing for the same reason. ForEach should really fail if there are no pods. Hopefully we haven't been masking broken tests for the past month.

**Comment by jlowdermilk** [comment]
Fixed by #25598
